python/lsst/daf/butler/_butlerRepoIndex.py

```diff
@@ -0,0 +1,173 @@
+    """Index of all known butler repositories.
+
+    The index of butler repositories is found by looking for a
+    configuration file at the URI pointed at by the environment
+    variable ``$BUTLER_REPOSITORY_INDEX``. The configuration file
+    is a simple dictionary lookup of the form:
+
+    .. code-block:: yaml
+
+       label1: uri1
+       label2: uri2
+
+    and can be in YAML or JSON format. The content of the file will be
+    cached.
+    """
+
+    index_env_var: ClassVar[str] = "BUTLER_REPOSITORY_INDEX"
+    """The name of the environment variable to read to locate the index."""
+
+    _cache: ClassVar[Dict[ButlerURI, Config]] = {}
+    """Cache of indexes. In most scenarios only one index will be found
+    and the environment will not change. In tests this may not be true."""
+
+    @classmethod
+    def _read_repository_index(cls, index_uri: ButlerURI) -> Config:
+        """Read the repository index from the supplied URI.
+
+        Parameters
+        ----------
+        index_uri : `str` or `ButlerURI`
+            URI of the repository index.
+
+        Returns
+        -------
+        repo_index : `Config`
+            The index found at this URI.
+
+        Raises
+        ------
+        FileNotFoundError
+            Raised if the URI does not exist.
+
+        Notes
+        -----
+        Does check the cache before reading the file.
+        """
+        if index_uri in cls._cache:
+            return cls._cache[index_uri]
+
+        repo_index = Config(index_uri)
+        cls._cache[index_uri] = repo_index
+
+        return repo_index
+
+    @classmethod
+    def _get_index_uri(cls) -> ButlerURI:
+        """Find the URI to the repository index.
+
+        Returns
+        -------
+        index_uri : `ButlerURI`
+            URI to the repository index.
+
+        Raises
+        ------
+        KeyError
+            Raised if the location of the index could not be determined.
+        """
+        index_uri = os.environ.get(cls.index_env_var)
+        if index_uri is None:
+            raise KeyError(f"No repository index defined in enviroment variable {cls.index_env_var}")
+        return ButlerURI(index_uri)
+
+    @classmethod
+    def _read_repository_index_from_environment(cls) -> Config:
+        """Look in environment for index location and read it.
+
+        Returns
+        -------
+        repo_index : `Config`
+            The index found in the environment.
+        """
+        index_uri = cls._get_index_uri()
+        return cls._read_repository_index(index_uri)
+
+    @classmethod
+    def get_known_repos(cls) -> Set[str]:
+        """Retrieve the list of known repository labels.
+
+        Returns
+        -------
+        repos : `set` of `str`
+            All the known labels. Can be empty if no index can be found.
+        """
+        try:
+            repo_index = cls._read_repository_index_from_environment()
+        except (FileNotFoundError, KeyError):
+            return set()
```

Comment:
It's not entirely clear to me that it's nicer to return an empty set than to pass along the exception.  It means that there's no available reason for knowing why the set is empty.

Reply:
I was wondering about that. On the one hand it's nice not to have to put it in a try block and it seems like no results is going to be a strong indicator that something odd is happening. On the other hand you can't then use that knowledge to work out what the actual problem is unless I add logging (or you do `get_repo_uri()` and look at that exception). Maybe some debug log messages wouldn't hurt anyhow.  If you really want it to raise I'll change it.

Reply:
It's a shame you can't return an empty set _and_ an error string :(.  And adding `get_known_repos_error()` doesn't seem reasonable.  I guess just leave it.